beacon_node/beacon_chain/src/shuffling_cache.rs

```diff
@@ -127,13 +140,55 @@ impl ShufflingCache {
             // worth two in the promise-bush!
             .map_or(true, CacheItem::is_promise)
         {
-            self.cache.put(
+            self.insert_cache_item(
                 key,
                 CacheItem::Committee(committee_cache.to_arc_committee_cache()),
             );
         }
     }
 
+    /// Prunes the cache first before inserting a new cache item.
+    fn insert_cache_item(&mut self, key: AttestationShufflingId, cache_item: CacheItem) {
+        self.prune_cache();
+        self.cache.insert(key, cache_item);
+    }
+
+    /// Prunes the `cache` to keep the size below the `cache_size` limit, based on the following
+    /// preferences:
+    /// - Entries from more recent epochs are preferred over older ones.
+    /// - Entries with shuffling ids matching the head's previous, current, and future epochs must
+    ///   not be pruned.
+    fn prune_cache(&mut self) {
+        if self.cache.len() >= self.cache_size {
+            let prune_count = self.cache.len() - self.cache_size + 1;
+            let shuffling_ids_to_prune = self
+                .cache
+                .keys()
+                .sorted_by_key(|key| key.shuffling_epoch)
+                .filter(|shuffling_id| {
+                    Some(shuffling_id)
+                        != self
+                            .head_shuffling_ids
+                            .id_for_epoch(shuffling_id.shuffling_epoch)
+                            .as_ref()
+                            .as_ref()
+                })
+                .take(prune_count)
+                .cloned()
+                .collect::<Vec<_>>();
+
+            for shuffling_id in shuffling_ids_to_prune.iter() {
+                self.cache.remove(shuffling_id);
+            }
+
+            debug!(
```

Comment:
Maybe we could push this up into the `for` loop above and log each shuffling ID that we remove? I expect we'd only remove one at a time and it might be nice when debugging to see which shufflings are being removed.

Reply:
Yep good idea! Updated in https://github.com/sigp/lighthouse/pull/4296/commits/db226fe9f58cc94040af011ecdd30068a598cdfb